docs/config-files.md

```diff
@@ -0,0 +1,228 @@
+
+
+## Config Format
+
+The format of individual config files themselves separates into JS files vs [JSON5](https://json5.org/) files.
+
+### JSON5
+
+Any file that isn't a `.js` file will be parsed as JSON5 and should contain an object matching
+the [options](options.md) format that Babel accepts.
+
+### JavaScript
+
+Any `.js` file will be `require()`ed and should export either a configuration object, or a function
+that will return a configuration object when called. The main benefit being that users can include
+JS logic to build up their config structures, potentially allowing config logic to be shared
+more easily. `.js` files can be used as [project-wide configuration](#project-wide-configuration) or
+via `.babelrc.js` files for [file-relative configuration](#file-relative-configuration).
+
+Function-returning configs are given a few special powers because they can access an API exposed
+by Babel itself. See [Config Function API](#config-function-api) for more information.
+
+## Config Function API
+
+JS config files may export a function that will be passed config function API:
+
+```js
+module.exports = function(api) {
+  return {};
+}
+```
+
+The `api` object exposes everthing Babel itself exposes from its index module, along with
+config-file specific APIs:
+
+### `api.version`
+
+Type: `string`<br />
+
+The version string for the Babel version that is loading the config file.
+
+### `api.cache`
+
+JS configs are great because they can compute a config on the fly, but the downside
+there is that it makes caching harder. Babel wants to avoid re-executing the
+config function every time a file is compiled, because then it would also need to
+re-execute any plugin and preset functions referenced in that config.
+
+The avoid this, Babel expects users of config functions to tell it how to manage
+caching within a config file.
+
+* `api.cache.forever()` - Permacache the computed config and never call the function again.
+* `api.cache.never()` - Do not cache this config, and re-execute the function every time.
+* `api.cache.using(() => process.env.NODE_ENV)` - Cache based on the value of `NODE_ENV`.
+  Any time the `using` callback returns a value other than the one that was expected, the overall
+  config function will be called again and a new entry will be added to the cache.
+* `api.cache.invalidate(() => process.env.NODE_ENV)` - Cache based on the value of `NODE_ENV`.
+  Any time the `using` callback returns a value other than the one that was expected, the overall
+  config function will be called again and all entries in the cache will be replaced with the result.
+
+Since the actual callback result is used to check if the cache entry is valid, it is recommended
+that:
+
+* Callbacks should be small and side-effect free.
+* Callbacks should return values with the smallest range possible. For example, the
+  `.using(() => process.env.NODE_ENV)` usage above is not idea because it would create an unknown
```

Comment:
Here is a typo, I think you want to write ideal

Reply:
Thanks!